flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/JoinUtil.scala

```diff
@@ -0,0 +1,468 @@
+      while (i < call.getOperands.size) {
+        val (curIsExistSysTimeAttr, curTimeType, curIsLeftStreamAttr) =
+          analyzeTimeExpression(call.getOperands.get(i), leftFieldCount, inputType)
+        if (isExistTimeIndicator && curIsExistSysTimeAttr) {
+          throw TableException(
+            s"Equality join time conditon can not include duplicate {$timeType} attribute."
+          )
+        }
+        if (curIsExistSysTimeAttr) {
+          isExistTimeIndicator = curIsExistSysTimeAttr
+          timeType = curTimeType
+          isLeftStreamAttr = curIsLeftStreamAttr
+        }
+
+        i += 1
+      }
+
+    }
+    (isExistTimeIndicator, timeType, isLeftStreamAttr)
+  }
+
+  /**
+    * Calcute the time boundary. Replace the rowtime/proctime with zero literal.
+    * such as:
+    *  a.proctime - inteval '1' second > b.proctime - interval '1' second - interval '2' second
+    *  |-----------left--------------|   |-------------------right---------------------------\
+    * then the boundary of a is right - left:
+    *  ((0 - 1000) - 2000) - (0 - 1000) = -2000
+    */
+  private def reduceTimeExpression(
+    leftNode: RexNode,
+    rightNode: RexNode,
+    rexBuilder: RexBuilder,
+    config: TableConfig): RexLiteral = {
+
+    val replLeft = replaceTimeIndicatorWithLiteral(leftNode, rexBuilder, true)
+    val replRight = replaceTimeIndicatorWithLiteral(rightNode, rexBuilder,false)
+    val literalRex = rexBuilder.makeCall(SqlStdOperatorTable.MINUS, replLeft, replRight)
+
+    val exprReducer = new ExpressionReducer(config)
+    val originList = new util.ArrayList[RexNode]()
+    originList.add(literalRex)
+    val reduceList = new util.ArrayList[RexNode]()
+    exprReducer.reduce(rexBuilder, originList, reduceList)
+
+    reduceList.get(0) match {
+      case call: RexCall => call.getOperands.get(0).asInstanceOf[RexLiteral]
+      case literal: RexLiteral => literal
+      case _ =>
+        throw TableException(
+          s"Equality join time condition only support constant."
+        )
+
+    }
+  }
+
+  /**
+    * replace the rowtime/proctime with zero literal.
+    * Because calculation between timestamp can only be TIMESTAMP +/- INTERVAL
+    * so such as b.proctime + interval '1' hour - a.proctime
+    * will be translate into TIMESTAMP(0) + interval '1' hour - interval '0' second
+    */
+  private def replaceTimeIndicatorWithLiteral(
+    expr: RexNode,
+    rexBuilder: RexBuilder,
+    isTimeStamp: Boolean): RexNode = {
+    if (expr.isInstanceOf[RexCall]) {
+      val call: RexCall = expr.asInstanceOf[RexCall]
+      var i = 0
+      val operands = new util.ArrayList[RexNode]
+      while (i < call.getOperands.size) {
+        val newRex =
+          replaceTimeIndicatorWithLiteral(call.getOperands.get(i), rexBuilder, isTimeStamp)
+        operands.add(newRex)
+        i += 1
+      }
+      rexBuilder.makeCall(call.getType, call.getOperator, operands)
+    } else if (expr.isInstanceOf[RexInputRef]) {
```

Comment:
Are we sure at this point that there are no other fields included in the condition?